std/os.zig

```diff
@@ -3509,20 +1866,566 @@ pub const MProtectError = error{
+    InvalidUtf8,
+
+    PathAlreadyExists,
+
+    Unexpected,
+};
+
+/// Return the canonicalized absolute pathname.
+/// Expands all symbolic links and resolves references to `.`, `..`, and
+/// extra `/` characters in `pathname`.
+/// The return value is a slice of `out_buffer`, but not necessarily from the beginning.
+/// See also `realpathC` and `realpathW`.
+pub fn realpath(pathname: []const u8, out_buffer: *[MAX_PATH_BYTES]u8) RealPathError![]u8 {
+    if (windows.is_the_target and !builtin.link_libc) {
+        const pathname_w = try windows.sliceToPrefixedFileW(pathname);
+        return realpathW(&pathname_w, out_buffer);
+    }
+    const pathname_c = try toPosixPath(pathname);
+    return realpathC(&pathname_c, out_buffer);
+}
+
+/// Same as `realpath` except `pathname` is null-terminated.
+pub fn realpathC(pathname: [*]const u8, out_buffer: *[MAX_PATH_BYTES]u8) RealPathError![]u8 {
+    if (windows.is_the_target and !builtin.link_libc) {
+        const pathname_w = try windows.cStrToPrefixedFileW(pathname);
+        return realpathW(&pathname_w, out_buffer);
+    }
+    if (linux.is_the_target and !builtin.link_libc) {
+        const fd = try openC(pathname, O_PATH | O_NONBLOCK | O_CLOEXEC, 0);
+        defer close(fd);
+
+        var procfs_buf: ["/proc/self/fd/-2147483648\x00".len]u8 = undefined;
+        const proc_path = std.fmt.bufPrint(procfs_buf[0..], "/proc/self/fd/{}\x00", fd) catch unreachable;
+
+        return readlinkC(proc_path.ptr, out_buffer);
+    }
+    const result_path = std.c.realpath(pathname, out_buffer) orelse switch (std.c._errno().*) {
+        EINVAL => unreachable,
+        EBADF => unreachable,
+        EFAULT => unreachable,
+        EACCES => return error.AccessDenied,
+        ENOENT => return error.FileNotFound,
+        ENOTSUP => return error.NotSupported,
+        ENOTDIR => return error.NotDir,
+        ENAMETOOLONG => return error.NameTooLong,
+        ELOOP => return error.SymLinkLoop,
+        EIO => return error.InputOutput,
+        else => |err| return unexpectedErrno(@intCast(usize, err)),
+    };
+    return mem.toSlice(u8, result_path);
+}
+
+/// Same as `realpath` except `pathname` is null-terminated and UTF16LE-encoded.
+pub fn realpathW(pathname: [*]const u16, out_buffer: *[MAX_PATH_BYTES]u8) RealPathError![]u8 {
+    const h_file = try windows.CreateFileW(
+        pathname,
+        windows.GENERIC_READ,
+        windows.FILE_SHARE_READ,
+        windows.OPEN_EXISTING,
+        windows.FILE_ATTRIBUTE_NORMAL,
+    );
+    defer windows.CloseHandle(h_file);
+
+    var wide_buf: [windows.PATH_MAX_WIDE]u16 = undefined;
+    const wide_len = try windows.GetFinalPathNameByHandleW(h_file, &wide_buf, wide_buf.len, windows.VOLUME_NAME_DOS);
+    assert(wide_len <= wide_buf.len);
+    const wide_slice = wide_len[0..wide_len];
+
+    // Windows returns \\?\ prepended to the path.
+    // We strip it to make this function consistent across platforms.
```

Comment:
I'm not sure I understand this comment. Why do we strip `\\?\`? that would be like stripping `/` from the start of unix paths.

Reply:
This is a like a magic way to pass a flag to the windows syscall. It makes it look at the path-name is a very different way.

Reply:
There are many different types of windows paths; the `\\?\` prefix is the way to say "I know what I'm doing", and allows you avoid things like windows blowing up on files named `CON`.
See https://googleprojectzero.blogspot.com/2016/02/the-definitive-guide-on-win32-to-nt.html?m=1 for more info